pycqed/instrument_drivers/meta_instrument/LutMans/ro_lutman.py

```diff
@@ -262,105 +324,152 @@ def __init__(self, name, num_res: int=1, **kw):
+
+        return hash != expected_hash
+    
+    def _update_expected_program_hash(self):
+        """
+        Updates the expected AWG sequencer program hash with the current
+        hash of the sequencer program. This is intended to be called after
+        setting the hash.
+        """
+        hash = self.AWG.get_instr().get('awgs_0_sequencer_program_crc32_hash')
+        self.set('_awgs_ro_sequencer_program_expected_hash', hash)
+
+    def load_waveform_onto_AWG_lookuptable(
+        self, wave_id: str, regenerate_waveforms: bool=False):
+        """
+        Load a waveform into the AWG.
 
+        Args:
+            wave_id: can be either the "name" of a waveform or
+                the integer codeword corresponding to a combination of
+                readout waveforms.
+            regenerate_waveforms (bool) : if True regenerates all waveforms
+        """
+        if regenerate_waveforms:
+            self.generate_standard_waveforms()
+
+        if wave_id not in self.LutMap().keys():
+            wave_id = get_wf_idx_from_name(wave_id, self.LutMap())
+
+        # Create the combined waveform
+        I_wave = np.array([])
+        Q_wave = np.array([])
+        for resonator in self.LutMap()[wave_id]['resonators']:
+            # Create the waveform name
+            wavename = self.pulse_type() + '_R' + str(resonator)
+            # adding new wave (not necessarily same length)
+            I_wave = add_waves_different_length(
+                I_wave, self._wave_dict[wavename][0])
+            Q_wave = add_waves_different_length(
+                Q_wave, self._wave_dict[wavename][1])
+
+        # clipping the waveform
+        I_wave = np.clip(I_wave, self._voltage_min, self._voltage_max)
+        Q_wave = np.clip(Q_wave, self._voltage_min, self._voltage_max)
+
+        # I and Q hardcoded to channels 1 and 2, codeword hardcoded to zero
+        # in 'single_pulse' mode. In this mode we only ever upload a single
+        # waveform to the instrument!
+        if self._mode == 'single_pulse':
+            # With this extra if-statement we prevent uploading to other waveforms
+            # than codeword 0 with the waveform selected by the user.
+            if self._single_pulse_name == self.LutMap()[wave_id]['name']:
+                self.AWG.get_instr().set('wave_ch1_cw000', I_wave)
+                self.AWG.get_instr().set('wave_ch2_cw000', Q_wave)
+        else:
+            self.AWG.get_instr().set('wave_ch1_cw{:03}'.format(wave_id), I_wave)
+            self.AWG.get_instr().set('wave_ch2_cw{:03}'.format(wave_id), Q_wave)
+
+    def load_waveforms_onto_AWG_lookuptable(
+            self, 
+            regenerate_waveforms: bool=True,
+            stop_start: bool = True,
+            force_load_sequencer_program: bool=False):
+        # Uploading the codeword program (again) is needed to if the user
+        # has changed the mode of the instrument.
+        if self._program_hash_differs() or force_load_sequencer_program:
```

Comment:
We are now using the logic in the LutMan for determining when we reload. I understood that this should now be achieved from the AWG (inside the UFHQA) itself.

Reply:
@AdriaanRol This does not actually trigger a reload or recompilation of the program. It merely sets the mode that the UHFQA operates in. The check is made here because of the different operating modes of the LutMan selected using the `load_single_pulse_sequence_onto_UHFQC` and `load_DIO_triggered_seequence_onto_UHFQC`.